mesonbuild/dependencies/base.py

```diff
@@ -495,31 +493,53 @@ def __init__(self, name, environment, kwargs, language=None):
         # stored in the pickled coredata and recovered.
         self.pkgbin = None
 
-        # When finding dependencies for cross-compiling, we don't care about
-        # the 'native' pkg-config
-        if self.want_cross:
-            if 'pkgconfig' not in environment.cross_info.config['binaries']:
-                if self.required:
-                    raise DependencyException('Pkg-config binary missing from cross file')
+        for_machine = MachineChoice.HOST \
+            if self.want_cross and environment.is_cross_build() \
+            else MachineChoice.BUILD
+
+        # Create a nested function for sake of early return
+        def search():
+            # Only search for the pkg-config for each machine the first time and
+            # store the result in the class definition
+            if PkgConfigDependency.class_pkgbin[for_machine] is None:
+                mlog.debug('Pkg-config binary for %s is not cached.' % for_machine)
+            else:
+                mlog.debug('Pkg-config binary for %s is cached.' % for_machine)
+                choice = PkgConfigDependency.class_pkgbin[for_machine]
+                assert choice is not None
+                return choice
+            # Lookup in cross or machine file.
+            bt = environment.binaries[for_machine]
+            potential_pkgpath = bt.lookup_entry('pkgconfig')
+            if potential_pkgpath is None:
+                mlog.debug('Pkg-config binary missing from cross or native file, or PKG_CONFIG undefined.')
             else:
-                potential_pkgbin = ExternalProgram.from_bin_list(
-                    environment.cross_info.config['binaries'], 'pkgconfig')
-                if potential_pkgbin.found():
-                    self.pkgbin = potential_pkgbin
+                mlog.debug('Pkg-config binary for %s specified from config file as %s.' % (for_machine, potential_pkgpath))
+                potential_pkgbin = ExternalProgram.from_entry('pkgconfig', potential_pkgpath)
+                if not potential_pkgbin.found():
+                    mlog.debug('Pkg-config %s for machine %s specified at %s but not found.'
+                        % (potential_pkgbin.name, potential_pkgbin.command, for_machine))
```

Comment:
__[Flake8]__ <sub>posted by Sider</sub>

[E128] continuation line under-indented for visual indent